doc: add 2.1.0 release doc updates (#8510)
nrfconnect/sdk-nrf: doc/nrf/releases/release-notes-2.1.0.rst

The following list includes the summary of the most relevant changes introduced in this release.

* Added :ref:`experimental support <software_maturity>` for the nRF7002 Wi-Fi 6 companion IC with samples.
  See :ref:`nrf7002dk_nrf5340` and :ref:`wifi_samples` for more information.
* Added experimental support for `Matter over Wi-Fi with nRF7002 <ug_matter_overview_architecture_integration>`.
* Feature-completed experimental support for Matter over Thread.
  It uses the :ref:`Thread v1.3 certified stack <thread_ug_supported_features_v13>` introduced in the |NCS| v2.0.2.
* Added experimental support for setting the TX power envelope with FEM power control for Bluetooth® LE and 802.15.4.
* Added experimental support for Microsoft Azure Embedded C SDK for the nRF91 Series.
* Added :ref:`support <software_maturity>` for enhanced power optimized device management and nRF Cloud Location Services through LwM2M/CoAP with AVSystem for the nRF91 Series.
* Added support for :ref:`nRF Cloud full modem FOTA <nrf9160_fota>` for the nRF91 Series.
* Added experimental support for the Nordic Distance Toolbox (NDT) on the nRF5340 SoC with samples.
* Added Find My support for the nRF5340 SoC.

> * Updated Trusted Firmware-M to v1.6, with additional flash and RAM overhead optimizations.

Reviewer: ```suggestion
* Updated Trusted Firmware-M to v1.6, with additional RAM overhead optimizations.
```
The Flash optimization was already done for 2.0.0 release

Author: This is in highlights, so I need to double-check with stakeholders.

Author: Confirmed and edited.

Reviewer: Update on stakeholders opinion?

Author: "So we did not optimize flash consumption since last release? If this is the case, lets remove it. But if not, I would say its worth mentioning".

    These new functions and take a pointer to the AT monitor entry.

* :ref:`modem_key_mgmt` library:

  * Fixed an issue that would cause the library to assert on an unhandled CME error when the AT command failed to be sent.

* :ref:`at_cmd_parser_readme` library:

  * Fixed an issue that would cause AT command responses like ``+CNCEC_EMM`` with underscore to be filtered out.

* :ref:`pdn_readme` library:

  * Added:

    * Support for setting multiple event callbacks for the default PDP context.
    * The :c:func:`pdn_default_ctx_cb_dereg` function to deregister a callback for the default PDP context.
    * The :c:func:`pdn_esm_strerror` function to retrieve a textual description of an ESM error reason.
      The function is compiled when :kconfig:option:`CONFIG_PDN_ESM_STRERROR` Kconfig option is enabled.

  * Updated:

    * The :c:func:`pdn_default_callback_set` function name to :c:func:`pdn_default_ctx_cb_reg`.
    * Automatic subscribtion to ``+CNEC=16`` and ``+CGEREP=1`` if the :ref:`lte_lc_readme` library is used to change the modem's functional mode.

  * Removed the ``CONFIG_PDN_CONTEXTS_MAX`` Kconfig option.
    The maximum number of PDP contexts is now dynamic.

> * :ref:`nrf_modem_lib_readme` library:

Reviewer 2: This now reads a bit strange (Modem library integration layer library)
Link should perhaps point to https://developer.nordicsemi.com/nRF_Connect_SDK/doc/latest/nrfxlib/nrf_modem/README.html instead of 
https://developer.nordicsemi.com/nRF_Connect_SDK/doc/latest/nrf/libraries/modem/nrf_modem_lib.html

at least for the "headline", and then maybe add a link to the integration layer below?
@lemrey, thoughts?

Author: I'll remove "library" after the link. We can't link to nrfxlib here, as these are release notes for the sdk-nrf part, not sdk-nrfxlib part, which has its own release notes.

The Zephyr fork in |NCS| (``sdk-zephyr``) contains all commits from the upstream Zephyr repository up to and including ``71ef669ea4a73495b255f27024bcd5d542bf038c``, plus some |NCS| specific additions.

For the list of upstream Zephyr commits (not including cherry-picked commits) incorporated into nRF Connect SDK since the most recent release, run the following command from the :file:`ncs/zephyr` repository (after running ``west update``):

.. code-block:: none

   git log --oneline 71ef669ea4 ^45ef0d2

For the list of |NCS| specific commits, including commits cherry-picked from upstream, run:

.. code-block:: none

   git log --oneline manifest-rev ^71ef669ea4

The current |NCS| main branch is based on revision ``71ef669ea4`` of Zephyr.

The following list summarizes the major additions specific to the |NCS|:

* Added Wi-Fi L2 layer.
  This is a fork of Zephyr's Wi-Fi L2 with the added support for the WPA supplicant.

zcbor
=====

* Updated the `zcbor`_ module from v0.4.0 to v0.5.1.
  Release notes for v0.5.0 and v0.5.1 are located at :file:`ncs/modules/lib/zcbor/`.
* Regenerated :ref:`lib_fmfu_fdev` code using zcbor v0.5.1.

> Trusted Firmware-M

Reviewer: Added:
    TF-M support for FPU Hard ABI.

Updated:
    Updated to TF-M 1.6.0.

Fixed:
    An issue with Thingy:91 v1.5.0 and lower version.
    An issue where TF-M used more RAM compared to SPM in the minimal configuration
    An issue with non-secure storage partitions in external flash.

Modem-related libraries and versions
====================================

.. list-table:: Modem-related libraries and versions
   :widths: 15 10
   :header-rows: 1

   * - Library name
     - Version information
   * - Modem library
     - `Changelog <Modem library changelog for v2.1.0_>`_
   * - LwM2M carrier library
     - `Changelog <LwM2M carrier library changelog for v2.1.0_>`_

Known issues
************

Known issues are only tracked for the latest official release.
See `known issues for nRF Connect SDK v2.1.0`_ for the list of issues valid for the latest release.

.. _ncs_release_notes_210_changelog:

Changelog
*********

The following sections provide detailed lists of changes by component.

Application development
=======================

* Added:

  * Information about :ref:`gs_supported_OS` to the :ref:`gs_recommended_versions` page.
  * Information about :ref:`app_build_output_files` on the :ref:`app_build_system` page.
  * Information about :ref:`gs_debugging` on the :ref:`gs_testing` page.
    Also added links to this section in different areas of documentation.

* Updated:

  * :ref:`app_memory` page with a section about Gazell.

>   * :ref:`app_memory` page with a section about NFC.

Reviewer 3: better together ;)

  * New #XGPSDEL command to delete GNSS data from non-volatile memory.
  * New #XDFUSIZE command to get the size of the DFU file image.

* Updated:

  * The application to use :ref:`TF-M <ug_tfm>` enabled by default.
  * The GNSS service to signify location info to nRF Cloud.
  * The AT response and the URC sent when the application enters and exits data mode.
  * ``WAKEUP_PIN`` and ``INTERFACE_PIN`` to be now defined as *Active Low*.
    Both are *High* when the SLM application starts.

* Removed the software toggle of ``INDICATE_PIN`` in case of reset.

nRF5340 Audio
-------------

* Added:

  * Support for Basic Audio Profile, including support for the stereo BIS (Broadcast Isochronous Stream).
  * Bonding between gateway and headsets in the CIS (Connected Isochronous Stream).
  * DFU support for internal and external flash layouts.
    See :ref:`nrf53_audio_app_configuration_configure_fota` in the application documentation for details.
  * DFU advertising name based on role.

* Updated:

  * Network controller.
  * Documentation in the :ref:`nrf53_audio_app_building_script` section.
    The text now mentions how to recover the device if programming using script fails.

>   * Documentation of the operating temperature maximum range in the
>     :ref:`nrf53_audio_app_dk_features` and :ref:`nrf53_audio_app_dk_legal` sections.

Reviewer 4: ```suggestion
  * Documentation of the operating temperature maximum range in the :ref:`nrf53_audio_app_dk_features` and :ref:`nrf53_audio_app_dk_legal` sections.
```

Little tweak

* boot_serial:

  * Upgraded from cddl-gen v0.1.0 to zcbor v0.4.0.
  * Refactored and optimized the code, mainly in what affects the progressive erase implementation.
  * Fixed a compilation issue with the echo command code.

* imgtool: Added support for providing signature through a third party.
* zephyr:

  * Added initial support for leveraging the RAM-LOAD mode with the zephyr-rtos port.
  * Added the MCUboot status callback support.
    See :kconfig:option:`CONFIG_MCUBOOT_ACTION_HOOKS`.
  * Edited includes to have the ``zephyr/`` prefix.
  * Edited the DFU detection's GPIO-pin configuration to be done through DTS using the ``mcuboot-button0`` pin alias.
  * Edited the LED usage to prefer DTS' ``mcuboot-led0`` alias over the ``bootloader-led0`` alias.
  * Removed :c:func:`device_get_binding()` usage in favor of :c:func:`DEVICE_DT_GET()`.

* Documentation:

  * :ref:`mcuboot:readme-ncs` now includes information regarding the bootloader user guides in the |NCS| documentation.
  * Removed:

    * The Zephyr Test Plan page in the MCUboot documentation set.

>     * The Building and using MCUboot with Zephyr page in the MCUboot documentation set.

Reviewer 4: Could we highlight in italics the title of the old pages? It is not that clear right now

nRF9160: Asset Tracker v2
-------------------------

* Added:

  * :ref:`motion_impact_detection` using the ADXL372 accelerometer.
  * The following Kconfig options to set the threshold and timeout values:

      * :ref:`CONFIG_DATA_ACCELEROMETER_ACT_THRESHOLD <CONFIG_DATA_ACCELEROMETER_ACT_THRESHOLD>`
      * :ref:`CONFIG_DATA_ACCELEROMETER_INACT_THRESHOLD <CONFIG_DATA_ACCELEROMETER_INACT_THRESHOLD>`
      * :ref:`CONFIG_DATA_ACCELEROMETER_INACT_TIMEOUT_SECONDS <CONFIG_DATA_ACCELEROMETER_INACT_TIMEOUT_SECONDS>`

  * Support for full modem FOTA updates for nRF Cloud builds.

* Updated:

  * The application to use :ref:`TF-M <ug_tfm>` enabled by default.
  * Data sampling method.
    It is now performed when the device detects both activity and inactivity in passive mode, notified by the :c:enum:`SENSOR_EVT_MOVEMENT_INACTIVITY_DETECTED` event of the :ref:`sensor module <asset_tracker_v2_sensor_module>`.
  * ``CONFIG_MODEM_NEIGHBOR_SEARCH_TYPE`` Kconfig option.
  * The default value of the GNSS timeout in the application's :ref:`Real-time configurations <real_time_configs>`.

>     It is now set to 30 seconds.

Reviewer 5: I believe this overlaps with Line 253. Adding the 30 seconds as the default value there could allow removing these lines.

    To revert to NMEA, set the :ref:`CONFIG_GNSS_MODULE_NMEA <CONFIG_GNSS_MODULE_NMEA>` Kconfig option.
  * Forwarding of :c:enum:`SENSOR_EVT_MOVEMENT_ACTIVITY_DETECTED` and :c:enum:`SENSOR_EVT_MOVEMENT_INACTIVITY_DETECTED` events by the sensor module.
  * :ref:`Real-time configurations <real_time_configs>`, which can now configure the upper and lower thresholds for motion detection.
    You can also configure the timeout after which the sensor reports inactivity.
    The following Kconfig options set the threshold and timeout values:

    * :ref:`CONFIG_DATA_ACCELEROMETER_ACT_THRESHOLD <CONFIG_DATA_ACCELEROMETER_ACT_THRESHOLD>`
    * :ref:`CONFIG_DATA_ACCELEROMETER_INACT_THRESHOLD <CONFIG_DATA_ACCELEROMETER_INACT_THRESHOLD>`
    * :ref:`CONFIG_DATA_ACCELEROMETER_INACT_TIMEOUT_SECONDS <CONFIG_DATA_ACCELEROMETER_INACT_TIMEOUT_SECONDS>`

  * LwM2M schema.
    To use the new parameters, upload the updated ``config_object_descript.xml`` to AVSystem.
  * The conversions of RSRP and RSRQ.
    These now use common macros that follow the conversion algorithms defined in the `AT Commands Reference Guide`_.
  * Bootstrapping to be disabled by default.

>     This allows to connect to the default LwM2M service AVSystem's `Coiote Device Management`_ using free tier accounts.

Reviewer 5: ```suggestion
    This allows connecting to the default LwM2M service AVSystem's `Coiote Device Management`_ using free tier accounts.
```

  * ``CONFIG_DATA_ACCELEROMETER_THRESHOLD`` Kconfig option.
  * ``CONFIG_DATA_ACCELEROMETER_BUFFER_STORE`` Kconfig option.
  * ``CONFIG_DATA_ACCELEROMETER_BUFFER_COUNT`` Kconfig option.
  * ``SENSOR_EVT_MOVEMENT_DATA_READY`` event.

nRF9160: Serial LTE modem
-------------------------

* Added:

  * URC for GNSS timeout sleep event.
  * Selected flags support in #XRECV and #XRECVFROM commands.
  * Multi-PDN support in the Socket service.
  * New #XGPSDEL command to delete GNSS data from non-volatile memory.
  * New #XDFUSIZE command to get the size of the DFU file image.

* Updated:

  * The application to use :ref:`TF-M <ug_tfm>` enabled by default.
  * The GNSS service to signify location info to nRF Cloud.
  * The AT response and the URC sent when the application enters and exits data mode.
  * ``WAKEUP_PIN`` and ``INTERFACE_PIN`` to be now defined as *Active Low*.
    Both are *High* when the SLM application starts.

* Removed the software toggle of ``INDICATE_PIN`` in case of reset.

nRF5340 Audio
-------------

* Added:

  * Support for Basic Audio Profile, including support for the stereo BIS (Broadcast Isochronous Stream).

>   * Bonding between gateway and headsets in the CIS (Connected Isochronous Stream).

Reviewer 5: Shouldn't the acronyms be the ones in parenthesis, not the full term?

    The feature can be turned on using :kconfig:option:`CONFIG_CAF_BLE_STATE_SECURITY_REQ`.
  * nRF Desktop dongles to start peripheral discovery immediately after Bluetooth LE connection is established.
    The dongles no longer wait until the connection is secured.
  * Updated Bluetooth advertising data and scan response data logic:

    * The TX power included in the advertising packet is no longer hardcoded, the application reads it from the Bluetooth controller.
      The TX power is included in advertising packets even if the Bluetooth local identity in use has bond.
    * The UUID16 of Battery Service (BAS) and Human Interface Device Service (HIDS) are included in advertising packets only if the Bluetooth local identity in use has no bond.

Connectivity Bridge
-------------------

* Fixed:

  * Missing return statement that caused immediate asserts when asserts were enabled.
  * Too low UART RX timeout that caused high level of fragmentation of UART RX data.

Samples
=======

This section provides detailed lists of changes by :ref:`sample <sample>`, including protocol-related samples.
For lists of protocol-specific changes, see `Protocols`_.

> The following change applies to all relevant samples:

Reviewer 5: ```suggestion
The following changes apply to all relevant samples:
```

  * Removed the ``CONFIG_PDN_CONTEXTS_MAX`` Kconfig option.
    The maximum number of PDP contexts is now dynamic.

* :ref:`nrf_modem_lib_readme` library:

  * Added:

    * :kconfig:option:`CONFIG_NRF_MODEM_LIB_TRACE` Kconfig option that replaces :kconfig:option:`CONFIG_NRF_MODEM_LIB_TRACE_ENABLED`.
      The Kconfig option :kconfig:option:`CONFIG_NRF_MODEM_LIB_TRACE_ENABLED` is now deprecated and will be removed in the future.
    * A section about :ref:`modem_trace_backend_uart_custom_board` on the :ref:`nrf_modem_lib_readme` page.
    * :kconfig:option:`CONFIG_NRF_MODEM_LIB_MEM_DIAG` option to enable the :c:func:`nrf_modem_lib_diag_stats_get` function that retrieves memory runtime statistics, replacing the ``nrf_modem_lib_heap_diagnose`` and ``nrf_modem_lib_shm_tx_diagnose`` functions.

  * Updated:

    * Ability to add :ref:`custom trace backends <adding_custom_modem_trace_backends>`.
    * The trace module to use the new APIs in the modem library.
      The modem trace output is now handled by a dedicated thread that starts automatically.
      The trace thread is synchronized with the initialization and shutdown operations of the Modem library.

>     * Updated the following Kconfig options by refactoring them:

Reviewer 5: ```suggestion
    * The following Kconfig options by refactoring them:
```

      * ``CONFIG_NRF_MODEM_LIB_DEBUG_ALLOC`` and ``CONFIG_NRF_MODEM_LIB_DEBUG_SHM_TX_ALLOC`` into the new :kconfig:option:`CONFIG_NRF_MODEM_LIB_MEM_DIAG_ALLOC` Kconfig option.
      * ``CONFIG_NRF_MODEM_LIB_HEAP_DUMP_PERIODIC`` and ``CONFIG_NRF_MODEM_LIB_SHM_TX_DUMP_PERIODIC`` into the new :kconfig:option:`CONFIG_NRF_MODEM_LIB_MEM_DIAG_DUMP` Kconfig option.
      * ``CONFIG_NRF_MODEM_LIB_HEAP_DUMP_PERIOD_MS`` and ``CONFIG_NRF_MODEM_LIB_SHMEM_TX_DUMP_PERIOD_MS`` into the new :kconfig:option:`CONFIG_NRF_MODEM_LIB_MEM_DIAG_DUMP_PERIOD_MS` Kconfig option.

>     * Removed:
> 
>       * The following Kconfig options:
> 
>         * ``CONFIG_NRF_MODEM_LIB_TRACE_THREAD_PROCESSING``
>         * ``CONFIG_NRF_MODEM_LIB_TRACE_HEAP_SIZE``
>         * ``CONFIG_NRF_MODEM_LIB_TRACE_HEAP_SIZE_OVERRIDE``
>         * ``CONFIG_NRF_MODEM_LIB_TRACE_HEAP_DUMP_PERIODIC``
>         * ``CONFIG_NRF_MODEM_LIB_TRACE_HEAP_DUMP_PERIOD_MS``
>         * ``CONFIG_NRF_MODEM_LIB_DEBUG_ALLOC``
>         * ``CONFIG_NRF_MODEM_LIB_DEBUG_SHM_TX_ALLOC``
>         * ``CONFIG_NRF_MODEM_LIB_HEAP_DUMP_PERIODIC``
>         * ``CONFIG_NRF_MODEM_LIB_HEAP_DUMP_PERIOD_MS``
>         * ``CONFIG_NRF_MODEM_LIB_SHM_TX_DUMP_PERIODIC``
>         * ``CONFIG_NRF_MODEM_LIB_SHMEM_TX_DUMP_PERIOD_MS``
> 
>       * The following functions:
> 
>         * ``nrf_modem_lib_trace_start``
>         * ``nrf_modem_lib_trace_stop``
>         * ``nrf_modem_lib_heap_diagnose``
>         * ``nrf_modem_lib_shm_tx_diagnose``

Reviewer 5: ```suggestion
  * Removed:

    * The following Kconfig options:

      * ``CONFIG_NRF_MODEM_LIB_TRACE_THREAD_PROCESSING``
      * ``CONFIG_NRF_MODEM_LIB_TRACE_HEAP_SIZE``
      * ``CONFIG_NRF_MODEM_LIB_TRACE_HEAP_SIZE_OVERRIDE``
      * ``CONFIG_NRF_MODEM_LIB_TRACE_HEAP_DUMP_PERIODIC``
      * ``CONFIG_NRF_MODEM_LIB_TRACE_HEAP_DUMP_PERIOD_MS``
      * ``CONFIG_NRF_MODEM_LIB_DEBUG_ALLOC``
      * ``CONFIG_NRF_MODEM_LIB_DEBUG_SHM_TX_ALLOC``
      * ``CONFIG_NRF_MODEM_LIB_HEAP_DUMP_PERIODIC``
      * ``CONFIG_NRF_MODEM_LIB_HEAP_DUMP_PERIOD_MS``
      * ``CONFIG_NRF_MODEM_LIB_SHM_TX_DUMP_PERIODIC``
      * ``CONFIG_NRF_MODEM_LIB_SHMEM_TX_DUMP_PERIOD_MS``

    * The following functions:

      * ``nrf_modem_lib_trace_start``
      * ``nrf_modem_lib_trace_stop``
      * ``nrf_modem_lib_heap_diagnose``
      * ``nrf_modem_lib_shm_tx_diagnose``
```
Removed two spaces each, to fix indentation level.

    * Handling of timeout errors when using CoAP.

Other libraries
---------------

* Added:

  * Documentation for the :ref:`lib_adp536x` library.
  * Documentation for the :ref:`lib_flash_map_pm` library.

* :ref:`lib_flash_patch` library:

  * Added documentation page.
  * Modified the :kconfig:option:`CONFIG_DISABLE_FLASH_PATCH` Kconfig option, so that it can be used on the nRF52833 SoC.

* :ref:`doc_fw_info` module:

  * Fixed a bug where MCUboot would experience a fault when using the :ref:`doc_fw_info_ext_api` feature.

* :ref:`nrf_rpc_ipc_readme`:

  * This is a new library.
    It can use different transport implementation for each nRF RPC group.
  * Memory for remote procedure calls is now allocated on a heap instead of the calling thread stack.

* :ref:`emds_readme`:

  * Updated :c:func:`emds_entry_add` to no longer use heap, but instead require a pointer to the dynamic entry structure :c:struct:`emds_dynamic_entry`.
    The dynamic entry structure should be allocated in advance.

* :ref:`mod_memfault`:

  * Added default metrics for Bluetooth.

Common Application Framework (CAF)
----------------------------------

* :ref:`caf_ble_adv`:

>   * Added :kconfig:option:`CONFIG_CAF_BLE_ADV_FILTER_ACCEPT_LIST` Kconfig option.
>     The option is used instead of :kconfig:option:`CONFIG_BT_FILTER_ACCEPT_LIST` option to enable the filter accept list.
>   * Integrated :ref:`bt_le_adv_prov_readme`.
>     The subsystem is now used instead of the :file:`*.def` file to configure advertising data and scan response data.
>   * Bluetooth device name is no longer automatically included in scan response data.
>     A dedicated data provider (:kconfig:option:`CONFIG_BT_ADV_PROV_DEVICE_NAME`) can be used to add the Bluetooth device name to the scan response data.
>   * Added :c:struct:`ble_adv_data_update_event` that can be used to trigger update of advertising data and scan response data during undirected advertising.
>     When the event is received, the module gets new data from providers and updates advertising payload.
>   * Added a wakeup call when connection is made in the grace period.
>     With this change, the call wakes up the whole system to avoid inconsistent power state between modules.

Reviewer 5: ```suggestion
  * Added:
    *  :kconfig:option:`CONFIG_CAF_BLE_ADV_FILTER_ACCEPT_LIST` Kconfig option.
      The option is used instead of :kconfig:option:`CONFIG_BT_FILTER_ACCEPT_LIST` option to enable the filter accept list.
      * :c:struct:`ble_adv_data_update_event` that can be used to trigger update of advertising data and scan response data during undirected advertising.
      When the event is received, the module gets new data from providers and updates advertising payload.
    * A wakeup call when connection is made in the grace period.
      With this change, the call wakes up the whole system to avoid inconsistent power state between modules.

  * Integrated :ref:`bt_le_adv_prov_readme`.
    The subsystem is now used instead of the :file:`*.def` file to configure advertising data and scan response data.
  * Bluetooth device name is no longer automatically included in scan response data.
    A dedicated data provider (:kconfig:option:`CONFIG_BT_ADV_PROV_DEVICE_NAME`) can be used to add the Bluetooth device name to the scan response data.
  
```

The following list summarizes the most important changes inherited from the upstream Matter:

* Added:

  * Support for Matter device factory data.
    This includes a set of scripts for building the factory data partition content, and the ``FactoryDataProvider`` class for accessing this data.
  * :ref:`Experimental support <software_maturity>` for Matter over Wi-Fi.

Thread
------

* Added information about Synchronized Sleepy End Device (SSED) and SED vs SSED activity in the :ref:`thread_ot_device_types` documentation.
* Removed multiprotocol support from :file:`overlay-cert.config` and moved it to :file:`overlay-multiprotocol.conf`.

See `Thread samples`_ for the list of changes for the Thread samples.

Zigbee
------

* Updated:

  * The PAN ID conflict resolution is now enabled in applications that use the :ref:`lib_zigbee_application_utilities` library.
    For details, see `Libraries for Zigbee`_.
  * The default entropy source of Zigbee samples and unit tests to Cryptocell for SoCs that have Cryptocell.

See `Zigbee samples`_ for the list of changes for the Zigbee samples.

> 

Reviewer 6: Please add a HomeKit section:

HomeKit
----------
* Added:

  * Support for Thread 1.3.

* Updated:
  * OTA DFU using the iOS Home app (over UARP - BLE and Thread) is no longer experimental.
  * Integrated HomeKit Accessory Development Kit v6.3 (ADK v6.3  e6e82026).
  * Improved current consumption for Thread Sleepy End Devices (SEDs) and Bluetooth LE peripherals.
  * Improved documentation.
 
* Fixed:

  * An issue where Bluetooth LE TX configuration was set to 0 dBm by default.
  * An issue where the Stateless Switch application crashed upon factory reset.

Author: Added.

ESB
---

* Fixed the :c:func:`update_radio_crc` function in order to correctly configure the CRC's registers (8 bits, 16 bits, or none).

nRF IEEE 802.15.4 radio driver
------------------------------

* Added option to configure IEEE 802.15.4 ACK frame timeout at build time using :kconfig:option:`CONFIG_NRF_802154_ACK_TIMEOUT_CUSTOM_US`.

Wi-Fi
-----

* Added support for the nRF7002 DK that supports Wi-Fi SoC.
  For more information, see :ref:`ug_nrf70`.

See `Wi-Fi samples`_ for details about how to use Wi-Fi in your application.

RF Front-End Modules
====================

* Fixed a build error that occurred when building an application for nRF53 SoCs with Simple GPIO Front-End Module support enabled.
* Added the :kconfig:option:`CONFIG_MPSL_FEM_ONLY` Kconfig option that allows the :ref:`nrfxlib:mpsl_fem` API to be used without other MPSL features.
  The :ref:`MPSL library <nrfxlib:mpsl>` is linked into the build without initialization.
  You cannot use other MPSL features when this option is enabled.
* Added the possibility to add custom models which split the requested TX power between the power on SoC output and the FEM gain in a way desired by the user.

> Secure Partition Manager (SPM)

Reviewer 7: Should this entire section rather be moved under Libraries/Other libraries?

    * :c:func:`nrf_cloud_pgps_process_update` function that stores a portion of a P-GPS download to flash.
    * :c:func:`nrf_cloud_pgps_finish_update` function that a user of the P-GPS library calls when the custom download completes.

  * Reduced logging level for many messages to debug (``DBG``).

* :ref:`lib_azure_iot_hub` library:

  * Reworked to use `Azure SDK for Embedded C`_.
  * Updated the APIs for both IoT Hub and DPS interaction.
    The applications and samples that use the library have been updated accordingly.

* :ref:`lib_download_client` library:

  * Fixed:

    * Handling of duplicated CoAP packets.
    * Handling of timeout errors when using CoAP.

Other libraries
---------------

* Added:

  * Documentation for the :ref:`lib_adp536x` library.
  * Documentation for the :ref:`lib_flash_map_pm` library.
  * :ref:`lib_identity_key` library.

* :ref:`lib_flash_patch` library:

  * Added documentation page.
  * Modified the :kconfig:option:`CONFIG_DISABLE_FLASH_PATCH` Kconfig option, so that it can be used on the nRF52833 SoC.

* :ref:`doc_fw_info` module:

  * Fixed a bug where MCUboot would experience a fault when using the :ref:`doc_fw_info_ext_api` feature.

> * :ref:`nrf_rpc_ipc_readme`:

Author: Right, and it should not list changes, as it is new.

  * This is a new library.
    It can use different transport implementation for each nRF RPC group.

>   * Memory for remote procedure calls is now allocated on a heap instead of the calling thread stack.

Author: @KAGA164 , @peknis , this library went in in https://github.com/nrfconnect/sdk-nrf/pull/7510 . Two questions:

1. Are there going to be more libraries for nRF RPC? If not, the library should be moved to Other libraries and 1-entry section should be removed.
2. Is this a brand new library? If so, there is no point in listing fixes to this library in the release notes where it was introduced.

Reviewer 8: This is a single library only we only plan to implement different transport layers for it.
It is not brand new library, we released it in the NCS v1.8 probably

    This includes a set of scripts for building the factory data partition content, and the ``FactoryDataProvider`` class for accessing this data.
  * :ref:`Experimental support <software_maturity>` for Matter over Wi-Fi.

Thread
------

* Added information about Synchronized Sleepy End Device (SSED) and SED vs SSED activity in the :ref:`thread_ot_device_types` documentation.
* Removed multiprotocol support from :file:`overlay-cert.config` and moved it to :file:`overlay-multiprotocol.conf`.

See `Thread samples`_ for the list of changes for the Thread samples.

Zigbee
------

* Updated:

  * The PAN ID conflict resolution is now enabled in applications that use the :ref:`lib_zigbee_application_utilities` library.
    For details, see `Libraries for Zigbee`_.
  * The default entropy source of Zigbee samples and unit tests to Cryptocell for SoCs that have Cryptocell.

See `Zigbee samples`_ for the list of changes for the Zigbee samples.

HomeKit
-------

* Added support for Thread v1.3.
* Updated:

  * OTA DFU using the iOS Home app (over UARP - BLE and Thread).
    This feature is no longer experimental.
  * HomeKit Accessory Development Kit to v6.3 (ADK v6.3 e6e82026).

>   * Current consumption for Thread Sleepy End Devices (SEDs) and Bluetooth LE peripherals.

Reviewer 9: Reduced or improved current consumption

Author: Updated.

    This enables the radio peripheral's Direction Finding Extension for antenna switching.

* :ref:`bluetooth_direction_finding_connectionless_rx` sample:

  * Added devicetree overlay file for the nRF5340 application core that configures GPIO pin forwarding.
    This enables the radio peripheral's Direction Finding Extension for antenna switching.

* :ref:`bluetooth_direction_finding_connectionless_tx` sample:

  * Added devicetree overlay file for the nRF5340 application core that configures GPIO pin forwarding.
    This enables the radio peripheral's Direction Finding Extension for antenna switching.

* :ref:`bluetooth_direction_finding_peripheral` sample:

  * Added devicetree overlay file for the nRF5340 application core that configures GPIO pin forwarding.
    This enables the radio peripheral's Direction Finding Extension for antenna switching.

Bluetooth mesh samples
----------------------

* :ref:`bluetooth_mesh_light_lc` sample:

  * Added overlay file with support for storing data with :ref:`emds_readme`.
    Also changed the sample to restore Light state after power-down.

> nRF9160 samples

Reviewer 10: Perhaps here we could add a note about TF-M output being unavailable when Modem tracing is enabled?

Author: Right, forgot to add it. Let me do it now.

The code for integrating MCUboot into |NCS| is located in the :file:`ncs/nrf/modules/mcuboot` folder.

The following list summarizes both the main changes inherited from upstream MCUboot and the main changes specific to the |NCS|:

* boot_serial:

  * Upgraded from cddl-gen v0.1.0 to zcbor v0.4.0.
  * Refactored and optimized the code, mainly in what affects the progressive erase implementation.
  * Fixed a compilation issue with the echo command code.

* imgtool: Added support for providing signature through a third party.
* zephyr:

  * Added initial support for leveraging the RAM-LOAD mode with the zephyr-rtos port.
  * Added the MCUboot status callback support.
    See :kconfig:option:`CONFIG_MCUBOOT_ACTION_HOOKS`.
  * Edited includes to have the ``zephyr/`` prefix.
  * Edited the DFU detection's GPIO-pin configuration to be done through DTS using the ``mcuboot-button0`` pin alias.
  * Edited the LED usage to prefer DTS' ``mcuboot-led0`` alias over the ``bootloader-led0`` alias.
  * Removed :c:func:`device_get_binding()` usage in favor of :c:func:`DEVICE_DT_GET()`.

* Documentation:

>   * :ref:`mcuboot:mcuboot_ncs` now includes information regarding the bootloader user guides in the |NCS| documentation.

Reviewer 4: ```suggestion
  * Updated:

    * :ref:`mcuboot:mcuboot_ncs` now includes information regarding the bootloader user guides in the |NCS| documentation.
    * :ref:`ug_nrf5340` now includes information about using MCUboot's serial recovery of the network core image.
```

from this PR: https://github.com/nrfconnect/sdk-nrf/pull/8141

* Updated :ref:`app_memory` page with sections about Gazell and NFC.

RF Front-End Modules
--------------------

* Fixed a build error that occurred when building an application for nRF53 SoCs with Simple GPIO Front-End Module support enabled.
* Added the :kconfig:option:`CONFIG_MPSL_FEM_ONLY` Kconfig option that allows the :ref:`nrfxlib:mpsl_fem` API to be used without other MPSL features.
  The :ref:`MPSL library <nrfxlib:mpsl>` is linked into the build without initialization.
  You cannot use other MPSL features when this option is enabled.
* Added the possibility to add custom models which split the requested TX power between the power on SoC output and the FEM gain in a way desired by the user.

Wi-Fi
-----

* Added support for the nRF7002 DK that supports Wi-Fi SoC.
  For more information, see :ref:`ug_nrf70`.

See `Wi-Fi samples`_ for details about how to use Wi-Fi in your application.

Protocols
=========

This section provides detailed lists of changes by :ref:`protocol <protocols>`.
See `Samples`_ for lists of changes for the protocol-related samples.

Bluetooth LE
------------

> 

Reviewer 11: Please add "LTE dual-chip coex support on nRF52 is now ready for production"

Author: Updated.

* Added:

  * URC for GNSS timeout sleep event.
  * Selected flags support in #XRECV and #XRECVFROM commands.
  * Multi-PDN support in the Socket service.
  * New #XGPSDEL command to delete GNSS data from non-volatile memory.
  * New #XDFUSIZE command to get the size of the DFU file image.

* Updated:

  * The application to use :ref:`TF-M <ug_tfm>` enabled by default.
  * The GNSS service to signify location info to nRF Cloud.
  * The AT response and the URC sent when the application enters and exits data mode.
  * ``WAKEUP_PIN`` and ``INTERFACE_PIN`` to be now defined as *Active Low*.
    Both are *High* when the SLM application starts.

* Removed the software toggle of ``INDICATE_PIN`` in case of reset.

nRF5340 Audio
-------------

* Added:

  * :ref:`nrfxlib:lc3` module to the `sdk-nrfxlib`_.
    The software codec does not require additional configuration steps and special access anymore.
    This affects the configuration and building process of the application.
  * Support for Basic Audio Profile, including support for the stereo Broadcast Isochronous Stream (BIS).
  * Bonding between gateway and headsets in the Connected Isochronous Stream (CIS).

>   * DFU support for internal and external flash layouts.

Reviewer 12: Should be Experimental DFU support ...

Author: Updated, please check.

    This enables the radio peripheral's Direction Finding Extension for antenna switching.

* :ref:`bluetooth_direction_finding_peripheral` sample:

  * Added devicetree overlay file for the nRF5340 application core that configures GPIO pin forwarding.
    This enables the radio peripheral's Direction Finding Extension for antenna switching.

Bluetooth mesh samples
----------------------

* :ref:`bluetooth_mesh_light_lc` sample:

  * Added overlay file with support for storing data with :ref:`emds_readme`.
    Also changed the sample to restore Light state after power-down.

nRF9160 samples
---------------

* Added :ref:`modem_trace_backend_sample` sample, demonstrating how to add a custom modem trace backend.
  The custom backend prints the amount of trace data received in bytes, trace data throughput, and CPU load.
* Updated samples that support Thingy:91 to use :ref:`TF-M <ug_tfm>` enabled by default.

  .. note::
      A known issue was found that concerns :ref:`modem trace retrieval imcompatibility with TF-M (NCSDK-15512) <known_issues_other>`: You can either use **UART1** for TF-M output or for modem traces, but not for both.

>       This affects applications and samples based on nRF9160.

Reviewer 10: I'd move this right at the start of the nrf9160 samples section, since it's not related the "trace backend" or "thinghy:91" samples.

Reviewer 13: I agree.

  * Support for changing the default TX power using Kconfig options that start with ``CONFIG_BT_CTLR_TX_PWR_*``.

* Updated LTE dual-chip Coex support on the nRF52 Series.
  It is now ready for production.

For details, see the :ref:`SoftDevice Controller changelog <nrfxlib:softdevice_controller_changelog>`.

Bluetooth mesh
--------------

* Added support for using :ref:`emds_readme`.
  For details, see `Bluetooth mesh samples`_ and `Bluetooth libraries and services`_.
* Updated API in :ref:`bt_mesh_sensor_srv_readme`: Column get callback now gets called with a column index instead of a pointer to column.
  This was introduced to support a series for sensors with one or two channels.

Also see `Bluetooth mesh samples`_ for the list of changes.

Enhanced ShockBurst (ESB)
-------------------------

* Fixed the :c:func:`update_radio_crc` function in order to correctly configure the CRC's registers (8 bits, 16 bits, or none).

Matter
------

* Added support for Matter over Wi-Fi to several samples.
* Updated :ref:`ug_matter` with new pages about Matter SDK as well as information about Matter over Wi-Fi.

> * Removed the low-power configuration build type from all Matter samples.

Reviewer 14: Can we please rephrase it, as now it suggest we do not have low-power configuration.

In fact we have removed low-power configuration overlay's for few samples (Lock, Light Switch and Window Covering) and instead of requiring user to provide such overlay, we have enabled low-power communication modes by default.

  * Updated:

    * Usage of the :ref:`lib_modem_antenna` library to configure the GNSS antenna instead of configuring it directly.
    * :ref:`lib_nrf_cloud` library is no longer de-initialized and re-initialized on disconnect and reconnect.
    * The :ref:`lib_nrf_cloud` library's function :c:func:`nrf_cloud_gnss_msg_json_encode` is now used to send PVT location data instead of building an NMEA sentence.
    * Minor logging and function structure improvements.

* :ref:`nrf_cloud_rest_fota` sample:

  * Added support for full modem FOTA updates.

* :ref:`at_monitor_sample` sample:

  * Updated the conversions of RSRP and RSRQ.
    These now use common macros that follow the conversion algorithms defined in the `AT Commands Reference Guide`_.

Thread samples
--------------

* :ref:`ot_cli_sample` sample:

  * Added logging of errors and hard faults in CLI sample by default.
  * Updated the sample documentation with SRP information.

Matter samples
--------------

* Added optimized usage of the QSPI NOR flash sleep mode to reduce power consumption during the Matter commissioning.
* Updated the size of MCUBoot partition on ``nrf5340dk_nrf5340_cpuapp`` by reducing it by 16 kB.

> * Removed the low-power configuration build type from all Matter samples.

Reviewer 14: Seems to be duplication from above.

    * :kconfig:option:`CONFIG_PM_OVERRIDE_EXTERNAL_DRIVER_CHECK` to override the external driver check.
      This is needed when using an external flash which is not using the :ref:`QSPI NOR <zephyr:dtbinding_nordic_qspi_nor>` driver from Zephyr.

MCUboot
=======

The MCUboot fork in |NCS| (``sdk-mcuboot``) contains all commits from the upstream MCUboot repository up to and including ``1d4404116a9a6b54d54ea9aa3dd2575286e666cd``, plus some |NCS| specific additions.

The code for integrating MCUboot into |NCS| is located in the :file:`ncs/nrf/modules/mcuboot` folder.

The following list summarizes both the main changes inherited from upstream MCUboot and the main changes specific to the |NCS|:

* boot_serial:

  * Upgraded from cddl-gen v0.1.0 to zcbor v0.4.0.
  * Refactored and optimized the code, mainly in what affects the progressive erase implementation.
  * Fixed a compilation issue with the echo command code.

> * imgtool: Added support for providing signature through a third party.

Reviewer 15: alignment

* zephyr:

  * Added initial support for leveraging the RAM-LOAD mode with the zephyr-rtos port.
  * Added the MCUboot status callback support.
    See :kconfig:option:`CONFIG_MCUBOOT_ACTION_HOOKS`.
  * Edited includes to have the ``zephyr/`` prefix.
  * Edited the DFU detection's GPIO-pin configuration to be done through DTS using the ``mcuboot-button0`` pin alias.
  * Edited the LED usage to prefer DTS' ``mcuboot-led0`` alias over the ``bootloader-led0`` alias.
  * Removed :c:func:`device_get_binding()` usage in favor of :c:func:`DEVICE_DT_GET()`.

> 

Reviewer 15: worth to mention: Made possible serial recovery of the image of nRF5340 cpunet even that simultaneous cores upgrade is disabled (CONFIG_NRF_MULTI_IMAGE_UPDATE=n ). Ref CONFIG_NRF53_RECOVERY_NETWORK_CORE.
this was added in https://github.com/nrfconnect/sdk-nrf/pull/8050

Author: You want me to mention this in the zephyr additions section? 
Also: we mention this new section a couple of lines further below under doc changes.

  * Information about :ref:`gs_supported_OS` to the :ref:`gs_recommended_versions` page.
  * Information about :ref:`app_build_output_files` on the :ref:`app_build_system` page.
  * Information about :ref:`gs_debugging` on the :ref:`gs_testing` page.
    Also added links to this section in different areas of documentation.
  * An option to configure IEEE 802.15.4 ACK frame timeout at build time using :kconfig:option:`CONFIG_NRF_802154_ACK_TIMEOUT_CUSTOM_US`.

* Updated :ref:`app_memory` page with sections about Gazell and NFC.

RF Front-End Modules
--------------------

* Added:

  * The :kconfig:option:`CONFIG_MPSL_FEM_ONLY` Kconfig option that allows the :ref:`nrfxlib:mpsl_fem` API to be used without other MPSL features.
    The :ref:`MPSL library <nrfxlib:mpsl>` is linked into the build without initialization.
    You cannot use other MPSL features when this option is enabled.
  * The possibility to add custom models which split the requested TX power between the power on SoC output and the FEM gain in a way desired by the user.

* Fixed a build error that occurred when building an application for nRF53 SoCs with Simple GPIO Front-End Module support enabled.

Wi-Fi
-----

> * Added support for the nRF7002 DK that supports Wi-Fi SoC.

Reviewer 16: ```suggestion
* Added experimental support for the nRF7002 DK that includes the nRF7002 companion IC.
```

  * ``CONFIG_MODEM_CONVERT_RSRP_AND_RSPQ_TO_DB`` Kconfig option.
  * ``CONFIG_DATA_ACCELEROMETER_THRESHOLD`` Kconfig option.
  * ``CONFIG_DATA_ACCELEROMETER_BUFFER_STORE`` Kconfig option.
  * ``CONFIG_DATA_ACCELEROMETER_BUFFER_COUNT`` Kconfig option.
  * ``SENSOR_EVT_MOVEMENT_DATA_READY`` event.

nRF9160: Serial LTE modem
-------------------------

* Added:

  * URC for GNSS timeout sleep event.
  * Selected flags support in ``#XRECV`` and ``#XRECVFROM`` commands.
  * Multi-PDN support in the Socket service.
  * New ``#XGPSDEL`` command to delete GNSS data from non-volatile memory.
  * New ``#XDFUSIZE`` command to get the size of the DFU file image.

* Updated:

  * The application to use :ref:`TF-M <ug_tfm>` enabled by default.
  * The GNSS service to signify location information to nRF Cloud.
  * The AT response and the URC sent when the application enters and exits data mode.
  * ``WAKEUP_PIN`` and ``INTERFACE_PIN`` to be now defined as *Active Low*.
    Both are *High* when the SLM application starts.

* Fixed an issue where the features of the mbedTLS v3.1 were not enabled by default, which caused the native TLS to not work.

>   The documentation of the :ref:`SLM_AT_SOCKET`' socket option numbers was updated accordingly.

Author: @pirun , @FrancescoSer , please review and approve entry for #8707

Reviewer 17: Btw, the correct spelling is Mbed TLS.

  * The possibility to add custom models which split the requested TX power between the power on SoC output and the FEM gain in a way desired by the user.

* Fixed a build error that occurred when building an application for nRF53 SoCs with Simple GPIO Front-End Module support enabled.

Wi-Fi
-----

* Added experimental support for the nRF7002 DK that includes the nRF7002 companion IC.
  For more information, see the :ref:`ug_nrf70` guide.

See `Wi-Fi samples`_ for details about how to use Wi-Fi in your application.

Protocols
=========

This section provides detailed lists of changes by :ref:`protocol <protocols>`.
See `Samples`_ for lists of changes for the protocol-related samples.

Bluetooth LE
------------

* Added:

  * Support for changing the radio transmitter's default power level using the :c:func:`sdc_default_tx_power_set` function.
  * Support for changing the peripheral latency mode using the :c:func:`sdc_hci_cmd_vs_peripheral_latency_mode_set` function.
  * Support for changing the default TX power using Kconfig options that start with ``CONFIG_BT_CTLR_TX_PWR_*``.

* Updated LTE dual-chip Coex support on the nRF52 Series.

>   It is now ready for production.

Reviewer 18: Can we also get this into the highlight list, since it gets production supported?

Supported modem firmware
************************

See `Modem firmware compatibility matrix`_ for an overview of which modem firmware versions have been tested with this version of the |NCS|.

Use the latest version of the nRF Programmer app of `nRF Connect for Desktop`_ to update the modem firmware.
See :ref:`nrf9160_gs_updating_fw_modem` for instructions.

Modem-related libraries and versions
====================================

.. list-table:: Modem-related libraries and versions
   :widths: 15 10
   :header-rows: 1

   * - Library name
     - Version information
   * - Modem library
     - `Changelog <Modem library changelog for v2.1.0_>`_
   * - LwM2M carrier library
     - `Changelog <LwM2M carrier library changelog for v2.1.0_>`_

Known issues
************

Known issues are only tracked for the latest official release.
See `known issues for nRF Connect SDK v2.1.0`_ for the list of issues valid for the latest release.

.. _ncs_release_notes_210_changelog:

Changelog
*********

The following sections provide detailed lists of changes by component.

> * PENDING UPDATE

Reviewer 4: ```suggestion
* Updated:

  *  :ref:`mcuboot:readme-ncs.rst` now includes information regarding the bootloader user guides in the |NCS| documentation.

* Removed:

  * The *Zephyr Test Plan* page in the MCUboot documentation set.
  * The *Building and using MCUboot with Zephyr* page in the MCUboot documentation set.
```

If the ref to the mcuboot docset doesn't work (it might not be implemented), this is the direct link to that page:  https://developer.nordicsemi.com/nRF_Connect_SDK/doc/latest/mcuboot/readme-ncs.html

Author: Applied, please re-check @FrancescoSer .
